File: Python/_0516_Longest_Palindromic_Subsequence/Solution.py
# coding=utf-8
# Date: 2021/8/12 09:49


# 执行用时：4016 ms, 在所有 Python3 提交中击败了5.01%的用户
# 内存消耗：31.8 MB, 在所有 Python3 提交中击败了24.70%的用户
class Solution:
    def longestPalindromeSubseq(self, s: str) -> int:
        n = len(s)
        f = [[0 for _ in range(n)] for _ in range(n)]
        for length in range(1, n + 1):
            for i in range(n - (length - 1)):
                j = i + length - 1
                if length == 1:
                    f[i][j] = 1
                else:
                    if s[i] == s[j]:
                        f[i][j] = f[i + 1][j - 1] + 2
                    f[i][j] = max(f[i][j], max(f[i][j - 1], f[i + 1][j]))
        return f[0][n - 1]


if __name__ == "__main__":
    print(Solution().longestPalindromeSubseq("bbbab"))  # 4
    print(Solution().longestPalindromeSubseq("cbbd"))  # 2

File: Java/src/_0516_Longest_Palindromic_Subsequence/Solution.java
package _0516_Longest_Palindromic_Subsequence;

/**
 * Date: 2021/8/12 09:43
 * Content:
 * 执行用时：39 ms, 在所有 Java 提交中击败了62.85%的用户
 * 内存消耗：48.4 MB, 在所有 Java 提交中击败了54.86%的用户
 */
public class Solution {

    public int longestPalindromeSubseq(String cs) {
        char[] s = cs.toCharArray();
        int n = s.length;
        int[][] f = new int[n][n];
        for (int len = 1; len <= n; len++)
            for (int i = 0; i + len - 1 < n; i++) {
                int j = i + len - 1;
                if (len == 1) f[i][j] = 1;
                else {
                    if (s[i] == s[j]) f[i][j] = f[i + 1][j - 1] + 2;
                    f[i][j] = Math.max(f[i][j], Math.max(f[i][j - 1], f[i + 1][j]));
                }
            }
        return f[0][n - 1];
    }

    public static void main(String[] args) {

        System.out.println((new Solution()).longestPalindromeSubseq("bbbab"));  // 4
        System.out.println((new Solution()).longestPalindromeSubseq("cbbd"));  // 2
    }
}
